Change logging to use co-log-core instead of hslogger

michaelpj opened the pull request:
This started as an attempt to bubble up errors from the VFS as actual
errors and return them to the user via the LSP response. However, in
fact VFS operations occur in response to notifications, which don't have
responses.

So all we can do is log the error and drop the change, which is okay.
However, that made me look at how the logging works. At the moment we
use `hslogger`, which is fine, but isn't so great when it's plugging
into part of a larger system. For example, we might want to have a
global log handler that sends error-level logs to the client as
messages, or uses the `logMessage` method of the LSP spec. But there's no
way to intercept the messages sent by the VFS currently.

So I switched over to using `co-log-core`, which is also the direction
that [HLS is going](https://github.com/haskell/haskell-language-server/pull/2558).

`co-log-core` is also a lightweight dependency.
It's suboptimal for `lsp-types` to depend on a logging library, however, but that
should be fixed when we do https://github.com/haskell/lsp/issues/394.

jneira reviewed (approved):
approving by pure trust :-) so not very helpful, dont know who could do a useful review

michaelpj commented:
Yeah, I don't know who's active at the moment and would actually be interested in this. Actually, I wonder if @eddiemundo might have some thoughts, as the author of the HLS PR?

eddiemundo reviewed (approved):
There doesn't seem to any filtering on the priorities so you'll always see all messages, but if that's OK then it looks good to me.

michaelpj commented:
> There doesn't seem to any filtering on the priorities so you'll always see all messages, but if that's OK then it looks good to me.

Yeah, but only for the "easy" run method. For the "real" one you're expected to pass your own log action, so then you can filter as you like.

eddiemundo commented on lsp-test/src/Language/LSP/Test/Session.hs:
```diff
@@ -384,22 +384,24 @@ updateState (FromServerMess SWorkspaceApplyEdit r) = do
+  where
+        logger = L.LogAction $ \m -> case logSeverity m of { L.Error -> error $ show m; _ -> pure () }
+        checkIfNeedsOpened uri = do
           oldVFS <- vfs <$> get
 
           -- if its not open, open it
-          unless (toNormalizedUri uri `Map.member` vfsMap oldVFS) $ do
+          unless (has (vfsMap . ix (toNormalizedUri uri)) oldVFS) $ do
             let fp = fromJust $ uriToFilePath uri
             contents <- liftIO $ T.readFile fp
             let item = TextDocumentItem (filePathToUri fp) "" 0 contents
                 msg = NotificationMessage "2.0" STextDocumentDidOpen (DidOpenTextDocumentParams item)
             sendMessage msg
 
             modifyM $ \s -> do
-              let (newVFS,_) = openVFS (vfs s) msg
+              let newVFS = flip execState (vfs s) $ openVFS logger msg
               return $ s { vfs = newVFS }
```
Actually I think I get it, so the lsp doesn't actually do real logging except from tests and what not, we're just using the Error priority to signal an actual fail.

michaelpj replied:
This is in `lsp-test` which is for writing tests, and calls `error` all over the place if things don't go according to plan. This probably makes it a bit stricter than it was before, and maybe I should get rid of it, but I'm not 100% sure what to do with the logs here...

michaelpj replied:
Oh, and also this is the VFS logs, which aren't that important.

michaelpj commented:
Oh, something I should really add to the changelog: there were a couple of places where the lsp server itself was sending messages to the user. I changed those to instead log errors, and then it's up to the caller whether to use a log handler that sends errors to the user. And also we could export such a handler, which would be useful.

eddiemundo commented:
So by log handler you mean providing a `LogAction IO LspServerLog` that sends error logs happening in the process message loop to the client. Then an `lsp` implementer could pass that directly to `lsp`, or wrap it in their own `LogAction` if they want?
Yeah I agree it would maintain the old client visible behaviour as well as serve as an example of how to make a `LogAction`.

michaelpj commented:
Exactly. I think we've discussed in HLS having a log handler that sends _all_ messages with error severity to the client - this is then just a special case of that.

michaelpj commented:
Updated to use `WithSeverity` from `co-log`, and to provide some handlers for logging to the client via `window/logMessage` and `windowShowMessage`. However, you can't actually use such a handler without the server having started up properly, so this means unfortunately that running the server now needs *two* handlers: one for logging in `IO` before the server has started up fully, and one for logging in `LspM` that can be used after that and can send messages.

I updated the `example` server to use such a logger, which I think works quite nicely.

michaelpj commented:
I think this is getting to a good state, and I think we could merge it. There's some follow-ups to sort before we can release:
- We need a `co-log-core` release, but I think that's imminent
- This PR provides "nice" printing for the log messages through `Show`, but I would much rather use `Pretty` (which would probably mean writing instances for a bunch of types in `lsp-types`!)
- https://github.com/haskell/lsp/issues/394

I'll do some work on the latter two in parallel PRs, but they'll be a bit easier once this is merged.

michaelpj commented:
Issue for `Pretty` instances: https://github.com/haskell/lsp/issues/403

michaelpj commented:
Okay, I rebased on the VFS changes and just did local changes here to do `Pretty` instances for the logs. So the only todo is the `co-log-core` release.

michaelpj commented:
Okay, I'm going to merge this. I've built HLS with this and appropriate modifications, it works and logs correctly, so that seems okay.
